add scripts to download RedHat kernel headers

iurly opened the pull request:
translated Nathan's ruby script to Python and added a few improvements.
Notice the upload part is still untested as I'm awaiting write permission to artifactory bucket.

iurly committed:
add scripts to download RedHat kernel headers

nathan-b commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,198 @@
+
+#print(args)
+
+redhat_auth = RedHatTokenAuth(args.redhat_token)
```
The original script outputs a status line once authentication is complete. This is helpful for debugging.

iurly replied:
Good point. The thing is, this is now performed by a separate module which is not bound to the main script's logging (quite/verbose) logic. I'm thinking of using Python's `logging` module which would solve this problem without weird module intertwining.

nathan-b replied:
I still don't see an "authentication complete" status line here...is it somewhere else that I'm missing?

iurly replied:
It's currently a somehow transparent mechanism.
Actual authentication (i.e. "getting" aka "Refreshing" an access token) will only happen as soon the first API request is issued.
So, with -v (debugging) enabled, you'd see something like this:

```
[2021-09-08 16:38:58,943] INFO - Scraping bucket rhocp-4.7-for-rhel-8-x86_64-rpms
[2021-09-08 16:38:58,944] INFO - Refreshing RedHat access token
[2021-09-08 16:38:58,945] DEBUG - Starting new HTTPS connection (1): sso.redhat.com:443
[2021-09-08 16:38:59,216] DEBUG - https://sso.redhat.com:443 "POST /auth/realms/redhat-external/protocol/openid-connect/token HTTP/1.1" 200 1764
[2021-09-08 16:38:59,218] INFO - RedHat access token successfully refreshed
[2021-09-08 16:38:59,220] DEBUG - Starting new HTTPS connection (1): api.access.redhat.com:443
[2021-09-08 16:39:03,205] DEBUG - https://api.access.redhat.com:443 "GET /management/v1/packages/cset/rhocp-4.7-for-rhel-8-x86_64-rpms/arch/x86_64?limit=100&offset=0 HTTP/1.1" 200 None
```

The whole mechanism was inspired by HTTP Digest Authentication, where you need to receive a challenge from the server first, plus you might have per-thread state. But that's probably unnecessary in our case.
Are you worried about an invalid offline token not being detected until later?
In any case, it's addressed by the latest commit, adding pre-validation of the offline token.

nathan-b commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,198 @@
+        headers = {'accept': 'application/json'}
+        response = s.get(url, headers=headers, params=params, auth=redhat_auth)
+        response.raise_for_status()
+        #print (response.body)
```
Maybe print these when the `-v` switch is on?

iurly replied:
OK.

nathan-b commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,198 @@
+
+art_helper = None
+if args.artifactory_server and args.artifactory_key:
+    base_url = "https://{artifactory_server}/artifactory".format(artifactory_server=args.artifactory_server)
```
The artifactory server is not required to be at `/artifactory`, so the path is technically part of the artifactory server specifier.

iurly replied:
This was done for consistency with the existing `probe-builder` arguments. The whole parameter should be renamed `artifactory_base_url` then (as opposed to `artifactory_server`) and the same should be done in the `probe-builder`. @gnosek do you agree?

gnosek replied:
yup :+1:, just make sure you update the jenkins job that runs the probe builder one this gets merged

iurly replied:
sure, so we'll update the `probe-builder` to use the full base URL for consistency, right?

gnosek replied:
yes, that makes sense

nathan-b commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,198 @@
+            retries += 1
+            msg = "Encountered error {}: {}. ".format(j['error']['code'],{j['error']['message']})
+            if retries < MAX_RETRIES:
+                log (msg + "Attempting retry")
```
Inconsistent formatting (sometimes you write `log (` and sometimes `log(`. Doesn't python have a formatter that you can use to fix problems like these?

iurly replied:
Found `black`, thanks!

nathan-b commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,198 @@
+            pname = pkg_name(p)
+            #REMOVEME
+            #pname += "DELETEME." + pname
+            # add {pkg.rpm: url, sha256} to map
```
Remove these comments

nathan-b commented on download-rhel-headers/README.md:
```diff
@@ -0,0 +1,57 @@
+| `-o`, `--outdir` | Output directory (must exist) | `.` | `-o /tmp`, `-o headers` |
+| `-b`, `--redhat-bucket` | RedHat bucket(s) to walk (may be used multiple times) | `rhel-8-for-x86_64-baseos-rpms`, `rhel-8-for-x86_64-baseos-eus-rpms`, `hocp-4.7-for-rhel-8-x86_64-rpms` | `-b hocp-4.7-for-rhel-8-x86_64-rpms` |
+| `-a` | Artifactory server | (None) | `-a artifactory.mycompany.com` |
+| `-A` | Artifactory API key | (None) | `-A 998877aabbccdd` |
```
Is this just the API key, or the `username:api_key` tuple?

iurly replied:
Just the API key, as explained on line 31. But the descriptions/examples could use some improvements, I guess.

nathan-b commented on download-rhel-headers/redhat_auth.py:
```diff
@@ -0,0 +1,114 @@
+
+    def refresh_access_token(self):
+        s = requests.Session()
+        #print("\n***Refreshing access token!***")
```
This seems like it would be useful to log

iurly replied:
OK

nathan-b commented on download-rhel-headers/redhat_auth.py:
```diff
@@ -0,0 +1,114 @@
+            self._thread_local.num_401_calls = 1
+            return r
+
+        print("\n***Received 401!***")
```
Should this be done using `log()`?

iurly replied:
OK

nathan-b commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,198 @@
+            # We need to retry because sometimes the download server has a sulk
+            while num_retries < MAX_RETRIES:
+                try:
+                    resp = https_download_file(auth=redhat_auth, url = pkg['url'], ofname=fn, exp_sha256=pkg['sha256'])
```
If the specified output directory doesn't exist, the user doesn't find out until here. I recommend checking for existence before scanning the repository

iurly replied:
OK.

nathan-b commented on download-rhel-headers/redhat_auth.py:
```diff
@@ -0,0 +1,114 @@
+        self._thread_local.num_401_calls = 1
+        return r
+
+    def __eq__(self, other):
+        return all([self.offline_token == getattr(other, 'offline_token', None)])
+
+    def __ne__(self, other):
+        return not self == other
```
Does the library you use require these to be implemented?

iurly replied:
No idea. The existing `HTTPDigestAuth` does implement them, so I just assumed it wouldn't hurt to have them defined.

nathan-b reviewed (commented):
I tested the script and it seems to work great! A few suggestions for the code.

iurly committed:
reformat python code using black

iurly committed:
improve rhel download script

Address the following issues:
- add -i option (interactive mode)
  This will print succint progress reports on stderr
  (as opposed to full-fledged log lines which are streamed to stdout)
- use logging module across modules
- improved logging of redhat access token failure and refreshing steps
- add output directory existence check
- add artifactory argument consistency checking
- add preliminary artifactory credentials check (when provided)
  and cache artifact list to be reused across redhat buckets
- replace artifactory server (i.e. hostname) with full-fledged base URL
- general logging cleanup
- improved parameter examples in README.md

iurly committed:
README: minor cleanup, add sample output

iurly commented:
I should have addressed all the issues pointed out by @nathan-b.
The whole verbosity selection could probably use some improvement by allowing for explicit levels (warning, info, debug, etc...) as opposed to potentially-conflicting flags (i.e. both quiet and verbose).

nathan-b commented on download-rhel-headers/download_headers.py:
```diff
@@ -64,32 +60,64 @@ def https_download_file(auth, url, ofname, accept="application/octet-stream", ex
+    dest="buckets",
+    action="append",
+    default=[],
+    help="Redhat bucket(s) to scrape",
```
Red Hat

nathan-b commented on download-rhel-headers/download_headers.py:
```diff
@@ -64,32 +60,64 @@ def https_download_file(auth, url, ofname, accept="application/octet-stream", ex
+logging.basicConfig(
+    stream=sys.stdout,
+    level=level,
+    # format="[%(asctime)s] {%(filename)s:%(lineno)d} %(levelname)s - %(message)s",
```
Remove this line

nathan-b commented on download-rhel-headers/redhat_auth.py:
```diff
@@ -39,18 +40,13 @@ def refresh_access_token(self):
-        # print(js)
         self._thread_local.access_token = js["access_token"]
-        # print("Access token successfully refreshed!")
+        logging.info("RedHat access token successfully refreshed")
```
Red Hat

nathan-b commented on download-rhel-headers/download_headers.py:
```diff
@@ -64,32 +60,64 @@ def https_download_file(auth, url, ofname, accept="application/octet-stream", ex
+        base_url=args.artifactory_base_url, bucket=ARTIFACTORY_BUCKET, apikey=args.artifactory_key
+    )
+    # This has a double purpose:
+    # 1) pre-fetch the list of artifacst so it's only done once
```
artifacst -> artifacts

iurly replied:
.

nathan-b commented on download-rhel-headers/redhat_auth.py:
```diff
@@ -27,7 +28,7 @@ def init_per_thread_state(self):
     def refresh_access_token(self):
         s = requests.Session()
-        # print("\n***Refreshing access token!***")
+        logging.info("Refreshing RedHat access token")
```
Red Hat

nathan-b reviewed (dismissed):
This looks good! Just a few minor issues and you're ready to submit

iurly committed:
download-rhel-headers: cosmetic: fix typos

iurly committed:
pre-validate redhat offline token

Refresh (i.e. obtain) access token beforehand, so that
an invalid offline token is detected as early as possible.

gnosek commented on download-rhel-headers/README.md:
```diff
@@ -0,0 +1,105 @@
+
+| Parameter | Meaning | Default value | Examples |
+| --- | --- | --- | --- |
+| `-t` | Red Hat API token **(mandatory)** |  | `-t 6DhjYIFYM587anxFXg1XA6jTpNwWwoMuB2gBnlLR0SYYeGj4cpV1OAPmKaUfEDoh8XHIHWJvZ2D57bHdh5i7JOF7qedtphGNLXzGtRDTwyJqTYj5MScmqRy3VapnsgD2yhCBoJ1l4pJdvZuqblEZksOERPNGs5MKeBAMRazxLmlJF4ySv3tVhAp538KJLYtA7Z3We2fF6KKazdfLoaXy2uz0ViHiVidf7HHBwhsnlrBRjIqX4XJVrqeAEGnJvRb5dAjvOQDHxR0m7D3EQBA6AQEHu8ZCYPyRUL2JbBMi7DaSXZ3OdfSnfTiFprwF4TMnwbMEK9fuO3lMAaoBHA4WSmeSPFh ...
```
Is this a valid token? I hope not :D

gnosek commented on download-rhel-headers/README.md:
```diff
@@ -0,0 +1,105 @@
+| `-o`, `--outdir` | Output directory (must exist) | `.` | `-o /tmp`, `-o headers` |
+| `-b`, `--redhat-bucket` | RedHat bucket(s) to walk (may be used multiple times) | `rhel-8-for-x86_64-baseos-rpms`, `rhel-8-for-x86_64-baseos-eus-rpms`, `hocp-4.7-for-rhel-8-x86_64-rpms` | `-b hocp-4.7-for-rhel-8-x86_64-rpms` |
+| `-a` | Artifactory base url | (None) | `-a https://artifactory.mycompany.com/artifactory` |
+| `-A` | Artifactory API key | (None) | `-A  cmalzOJ7Vww8dKZv2Z7U9gifhjlrwQx6b44RPolzte00PEr9yVfjaGhclxGXXVFVre81S05Zs` |
```
same, I hope it's not a real token :D

iurly replied:
nope, it's the output from `head /dev/urandom | tr -dc A-Za-z0-9 | head -c73` :D

gnosek commented on download-rhel-headers/redhat_auth.py:
```diff
@@ -0,0 +1,119 @@
+    def __init__(self, offline_token):
+        self.offline_token = offline_token
+        # Keep state in per-thread local storage
+        self._thread_local = threading.local()
```
Why the thread local? I'd expect the state to be kept in the object itself, who cares if it gets moved to a different thread?

iurly replied:
it's legacy code from `HTTPDigestAuth`.
Ideally, I though we might want to obtain multiple access tokens (one per thread) to increase parallelism.
To be honest, it looks like the server will limit the number of outstanding requests based on something else (I reckon it's the IP address) so it might be useless here.

gnosek replied:
Would all these parallel downloads share the same auth object? It's surprising to see thread locals in such high level code IMHO

iurly replied:
I removed it (see latest commit)

gnosek commented on download-rhel-headers/redhat_auth.py:
```diff
@@ -0,0 +1,119 @@
+        if r.is_redirect:
+            self._thread_local.num_401_calls = 1
+
+    def handle_401(self, r, **kwargs):
```
The 401 counter isn't immediately obvious tbh. What's the idea behind it?

iurly replied:
You get 401 once -> access token expired -> get a new one and retry
You get 401 twice -> access token is invalid -> give up

gnosek replied:
Thanks! Can you make it a comment? :D

iurly replied:
Done

gnosek commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,252 @@
+                continue
+            else:
+                logging.critical(msg + "Aborting")
+                sys.exit(-2)
```
Technically the exit code is an unsigned 8-bit number

gnosek commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,252 @@
+    # 2) validate Artifactory credentials so to avoid a later failure
+    art_list = art_helper.list_artifacts()
+elif args.artifactory_base_url or args.artifactory_key:
+    sys.exit("Only one of --artifactory-base-url and --artifactory-key provided, but not both. Bailing out.")
```
umm. sys.exit("string")?

iurly replied:
Yup, https://docs.python.org/3/library/sys.html#sys.exit
Although the above doc states it should be 2 for command line issues so it's a good catch nonetheless.

gnosek replied:
TIL :) Thanks!

gnosek commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,252 @@
+    format="[%(asctime)s] %(levelname)s - %(message)s",
+)
+
+# Check if the output directory exists
+if not os.path.isdir(args.outdir):
+    sys.exit("Output directory '{}' ({}) does not exist. Aborting.".format(args.outdir, os.path.abspath(args.outdir)))
```
I'm wondering whether this is a better approach than just creating the directory:

```
try:
    os.makedirs(args.outdir)
except OSError as exc:
    if exc.errno != errno.EEXIST:
        raise
```

iurly replied:
We should ask @nathan-b if he has any objections against unsolicited directory creation

nathan-b replied:
Works for me. I think specifying a directory name as a parameter indicates sufficient intent to go ahead and create the directory or directories if not already present.

iurly replied:
Done with `FileExistsError` instead

gnosek commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,252 @@
+        headers = {"accept": "application/json"}
+        response = s.get(url, headers=headers, params=params, auth=redhat_auth)
+        response.raise_for_status()
+        # print (response.body)
+        j = response.json()
+        # print(j)
```
remove these comments?

gnosek commented on download-rhel-headers/download_headers.py:
```diff
@@ -0,0 +1,252 @@
+        # Determine whether we need to keep going or if we're at the end of the package list
+        pag = j["pagination"]
+        if not pag:
+            logging.critical("Got mal-formed JSON response:" + str(j))
```
```suggestion
            logging.critical("Got malformed JSON response:" + str(j))
```

iurly committed:
implement automatic directory creation, minor cleanups

iurly committed:
remove thread_local handling of access token

thread_local handling is legacy from HTTPDigestAuth where you
have sequential numbering of requests (i.e. nonce count),
which require special care when running in a threaded environment.

The concept was initially inherited for Red Hat authentication with the
idea that you could indefinitely speed up downloads by having multiple
threads, each with its own access token.

Since the actual number of parallel requests that you're allowed
to issue to Red Hat API servers is apparently limited to something like 4
per IP (or account/offline token, or subscription, or any other obscure
logic) and it's in no way influenced by the number of access tokens you're
using, this approach does not have any added value and just adds unneeded
complexity.

So remove it.

iurly committed:
README: add link for obtaining redhat token

iurly committed:
add support for RHEL 9 (beta)

- add two buckets (one for kernel-devel, one for kernel-core)
- fix regexp so to exclude kernel-devel-matched

iurly committed:
add support for multiple architectures (s390x)

iurly committed:
update rhel9 buckets

iurly committed:
rhel_headers: fix s390x build

iurly committed:
rhel_headers: add missing rhel-9-for-s390x-appstream-eus-rpms

iurly committed:
Merge remote-tracking branch 'origin/dev' into redhat_headers

iurly committed:
WIP: add redhat download to main builder entrypoint

nathan-b commented:
What's the status of this PR now?

iurly commented:
> What's the status of this PR now?

Right now? Open
One second from now? Closed
